File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemMoveByGravity.java
package com.example.try_gameengine.action;

import java.util.ArrayList;
import java.util.List;

import android.util.Log;

import com.example.try_gameengine.action.IGravityController.PathType;
import com.example.try_gameengine.action.MovementActionItemTrigger.MovementActionItemUpdateTimeDataDelegate;
import com.example.try_gameengine.action.info.MovementActionMoveByGravityInfo;
import com.example.try_gameengine.action.listener.IActionListener;
import com.example.try_gameengine.action.visitor.IMovementActionVisitor;

/**
 * MovementActionItemAlpha is a movement action that control alpha value.
 * @author irons
 *
 */
public class MovementActionItemMoveByGravity extends MovementActionItemUpdateTime implements Cloneable{ 
	
//	public MovementActionItemMoveByGravity(long millisTotal, long millisDelay, final int dx, final int dy){
//		this(millisTotal, millisDelay, dx, dy, "MovementItem");
//	}
//	
//	public MovementActionItemMoveByGravity(long millisTotal, long millisDelay, final int dx, final int dy, String description){
//		super(millisTotal, millisDelay, dx, dy, description);
//	}
	
	public MovementActionItemMoveByGravity(MovementActionMoveByGravityInfo info){
		super(info);
	}
	
	
	@Override
	public MovementAction getAction(){
		return this;
	}
	
	public List<MovementAction> getActions(){
		return actions;
	}

	@Override
	public MovementActionInfo getInfo() {
		// TODO Auto-generated method stub
		return info;
	}

	@Override
	public void setInfo(MovementActionInfo info) {
		// TODO Auto-generated method stub
		this.info = info;
	}
	
	@Override
	public List<MovementAction> getCurrentActionList() {
		// TODO Auto-generated method stub
		List<MovementAction> actions = new ArrayList<MovementAction>();
		actions.add(this);
		return actions;
	}
	
	@Override
	public List<MovementActionInfo> getCurrentInfoList() {
		// TODO Auto-generated method stub
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		currentInfoList.add(this.info);
		return infos;
	}
	
	@Override
	public List<MovementActionInfo> getMovementInfoList() {
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		return infos;
	}
	
	@Override
	public void cancelMove(){
		isStop = true;
		synchronized (MovementActionItemMoveByGravity.this) {
			MovementActionItemMoveByGravity.this.notifyAll();
		}
	}
	
	@Override
	void pause(){	
		data.setShouldPauseValue(data.getShouldActiveIntervalValue());
	}
	
	@Override
	public boolean isFinish(){
		return isStop;
	}
	
	public void setPathType(PathType pathType) {
		// TODO Auto-generated method stub
		((MovementActionMoveByGravityInfo)info).setPathType(pathType);
	}
	
	@Override
	public void accept(IMovementActionVisitor movementActionVisitor){
		movementActionVisitor.visitLeaf(this);
	}
	
	@Override
	protected MovementActionItemMoveByGravity clone() throws CloneNotSupportedException {
		MovementActionMoveByGravityInfo info = (MovementActionMoveByGravityInfo) getInfo().clone();
		MovementActionItemMoveByGravity movementActionItemMoveByGravity = new MovementActionItemMoveByGravity(info);
		return movementActionItemMoveByGravity;
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemMoveByGravity.kt
package com.example.try_gameengine.action

import com.example.try_gameengine.action.info.MovementActionMoveByGravityInfo
import com.example.try_gameengine.action.visitor.IMovementActionVisitor

/**
 * MovementActionItemAlpha is a movement action that control alpha value.
 * @author irons
 // */
class MovementActionItemMoveByGravity  //	public MovementActionItemMoveByGravity(long millisTotal, long millisDelay, final int dx, final int dy){
//		this(millisTotal, millisDelay, dx, dy, "MovementItem");
//	}
//	
//	public MovementActionItemMoveByGravity(long millisTotal, long millisDelay, final int dx, final int dy, String description){
//		super(millisTotal, millisDelay, dx, dy, description);
//	}
    (info: MovementActionMoveByGravityInfo?) : MovementActionItemUpdateTime(info), Cloneable {
    public override fun getAction(): MovementAction {
        return this
    }

    override fun getActions(): MutableList<MovementAction> {
        return actions
    }

    public override fun getInfo(): MovementActionInfo {
        // TODO Auto-generated method stub
        return info
    }

    public override fun setInfo(info: MovementActionInfo?) {
        // TODO Auto-generated method stub
        this.info = info ?: return
    }

    public override fun getCurrentActionList(): MutableList<MovementAction> {
        // TODO Auto-generated method stub
        val actions: MutableList<MovementAction> = ArrayList<MovementAction>()
        actions.add(this)
        return actions
    }

    public override fun getCurrentInfoList(): MutableList<MovementActionInfo?> {
        // TODO Auto-generated method stub
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        currentInfoList.add(this.info)
        return infos
    }

    public override fun getMovementInfoList(): MutableList<MovementActionInfo?> {
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        return infos
    }

    public override fun cancelMove() {
        isStop = true
        synchronized(this@MovementActionItemMoveByGravity) {
            (this@MovementActionItemMoveByGravity as Object).notifyAll()
        }
    }

    override fun pause() {
        data!!.setShouldPauseValue(data!!.getShouldActiveIntervalValue())
    }

    public override fun isFinish(): Boolean {
        return isStop
    }

    fun setPathType(pathType: IGravityController.PathType?) {
        // TODO Auto-generated method stub
        (info as MovementActionMoveByGravityInfo).setPathType(pathType)
    }

    public override fun accept(movementActionVisitor: IMovementActionVisitor) {
        movementActionVisitor.visitLeaf(this)
    }

    @Throws(CloneNotSupportedException::class)
    override fun clone(): MovementActionItemMoveByGravity {
        val info = getInfo()!!.clone() as MovementActionMoveByGravityInfo
        val movementActionItemMoveByGravity = MovementActionItemMoveByGravity(info)
        return movementActionItemMoveByGravity
    }
}
